common/src/ray_tracing.rs

```diff
@@ -24,48 +24,118 @@ pub fn trace_ray(
+            };
+        let local_ray = chunk.vertex.node_to_dual().cast::<f32>() * node_transform * ray;
+
+        let bounding_box = CubicVoxelRegion::from_ray_segment_and_radius(
+            dimension,
+            &local_ray,
+            result.tanh_distance,
+            chunk_ray_tracer.max_radius(),
         );
+
+        // Check collision within a single chunk
+        if let Some(bounding_box) = bounding_box {
+            chunk_ray_tracer.trace_ray(
+                &ChunkRayTracingContext {
+                    dimension,
+                    dimension_f32: dimension as f32,
+                    chunk,
+                    transform: math::mtranspose(&node_transform)
+                        * chunk.vertex.dual_to_node().cast(),
+                    voxel_data,
+                    ray: &local_ray,
+                    bounding_box,
+                },
+                &mut result,
+            );
+        }
+
+        // Compute the Klein-Beltrami coordinates of the ray segment's endpoints. To check whether neighboring chunks
+        // are needed, we need to check whether the endpoints of the line segments lie outside the boundaries of the square
+        // bounded by `klein_lower_boundary` and `klein_upper_boundary`.
+        let klein_ray_start = na::Point3::from_homogeneous(local_ray.position).unwrap();
+        let klein_ray_end =
+            na::Point3::from_homogeneous(local_ray.ray_point(result.tanh_distance)).unwrap();
+
+        // Add neighboring chunks as necessary, using one coordinate at a time.
+        for axis in 0..3 {
+            // Check for neighboring nodes
+            if klein_ray_start[axis] <= klein_lower_boundary
+                || klein_ray_end[axis] <= klein_lower_boundary
+            {
+                let side = chunk.vertex.canonical_sides()[axis];
+                let next_node_transform = side.reflection().cast::<f32>() * node_transform;
+                // Crude check to ensure that the neighboring chunk's node can be in the path of the ray. For simplicity, this
+                // check treats each node as a sphere and assumes the ray is pointed directly towards its center. The check is
+                // needed because chunk generation uses this approximation, and this check is not guaranteed to pass near corners.
+                let ray_node_distance = (next_node_transform * ray.position)
+                    .xyz()
+                    .magnitude()
+                    .acosh();
+                let ray_length = result.tanh_distance.atanh();
+                if ray_node_distance - ray_length
+                    > dodeca::BOUNDING_SPHERE_RADIUS as f32 + chunk_ray_tracer.max_radius()
+                {
+                    // Ray cannot intersect node
+                    continue;
+                }
+                // If we have to do collision checking on nodes that don't exist in the graph, we cannot have a conclusive result.
```

Comment:
Could we treat missing nodes/unpopulated chunks as solid walls for collision detection purposes?

Reply:
It's definitely possible, but it would be a compromise. Server-side, it's always a logic error if we collide with an unpopulated chunk, since we generate synchronously. This may be worth discussing pros and cons on Discord.